posthog/hogql/variables.py

```diff
@@ -0,0 +1,51 @@
+from typing import TypeVar
+
+from posthog.hogql import ast
+from posthog.hogql.errors import QueryError
+from posthog.hogql.visitor import CloningVisitor
+from posthog.models.insight_variable import InsightVariable
+from posthog.models.team.team import Team
+from posthog.schema import HogQLVariable
+
+T = TypeVar("T", bound=ast.Expr)
+
+
+def replace_variables(node: T, variables: list[HogQLVariable], team: Team) -> T:
+    return ReplaceVariables(variables, team).visit(node)
+
+
+class ReplaceVariables(CloningVisitor):
+    insight_variables: list[InsightVariable]
+
```

Comment:
What do you think about inlining this into `replace_placeholders`? First, every extra layer of `CloningVisitor` slows us down a bit, so they're best avoided. It's probably fine since there's an `if` blocking this from every query... but soon you might always have some global team-level variables just in case, it'll probably run on each query.

However the bigger reason for this suggestion is that the "[placeholder expressions](https://github.com/PostHog/posthog/pull/25216)" work is actually coming along nicely. We'll keep placeholders as a concept. Very soon I'd like all placeholders (filters, variables and regular placeholders) to be compiled into an object like `globals = { filters: ast.Expr(), 'filters.dateRange.from': ast.Expr(), variables: { 'key': 'value' }, ... }` and then fed directly to a `placeholder = execute_hog(compile_hog(placeholder.expr), globals)` line, [something like this](https://github.com/PostHog/posthog/pull/25040/files#diff-aecf9246a3b5a0113edc7b0e0603da97d761abd57277ebfe509c9ce1a1f0834fR44).

Reply:
I think combining these isn't the worst idea, but I'd like to do it once we've got the `globals` object. The logic between each of the `CloningVisitors` is pretty different at the moment, and that of variables is still evolving (unsure what it may look like in 2-3 PRs time, for example)